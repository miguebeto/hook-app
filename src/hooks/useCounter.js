import { useState } from "react";

export const useCounter = (initialstate = 10) => {
  //Creamos un estado con el valor inicial que será pasado por props a nuestro hook
  const [state, setState] = useState(initialstate); //10

  //Creamos la funcion seteadora que va incrementar el valor del state sumado al valor pasado por props, por defecto es 1
  const increment = ( factor = 1 ) => {
    setState( state + factor );
  };
  
  //Creamos la funcion seteadora que va incrementar el valor del state restando el valor pasado por props, por defecto es 1
  const decrement = ( factor = 1 ) => {
    setState( state - factor );
  };
  
  
  //Creamos la funcion seteadora que va a resetear el valor del estado a su estado iicial (10) pasado por parámetro 
  const reset = () => {
    setState( initialstate );
  };

  //retornamos el esatdo y las funciones seteadoras que necesitaremos usar 
  return {
    state, 
    increment, 
    decrement,
    reset
  };
};
