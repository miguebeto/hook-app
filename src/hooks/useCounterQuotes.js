import { useState } from "react";

export const useCounterQuotes = (initialstate = 10) => {

  //creamos un estado con initialstate que será recibido por props, por defecto es 10
  const [counter, setCounter] = useState(initialstate); //10

  //creamos las funciones que van a modificar el estado y luego serán exportadas
  const increment = () => {
   setCounter( counter + 1 );
  };

  const decrement = () => {
   setCounter( counter - 1 );
  };

  const reset = () => {
   setCounter( counter );
  };

  //Retomamos el estado y las funciones seteadoras
  return {
    counter, 
    increment, 
    decrement,
    reset
  };
};
