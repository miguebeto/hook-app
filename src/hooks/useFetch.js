import { useEffect, useRef, useState } from "react"; 

export const useFetch = (url) => {
  //Creamos una referencia con valor inicial en true
  const isMounted = useRef(true);

  //Creamos un esatdp con initial state de un objeto que tiene data, loading y error
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  //pasamos la referencia a false cuando se carga el componente 
  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  //setea el estado a loading true cada vez que se haga un peticion que cambia el loading por el resultado del fetch que se ejecuta cuando la url cambia, la url se pasa por props
  useEffect(() => {
    setState({ data: null, loading: true, error: null });

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setState({
            loading: false,
            error: null,
            data,
          });
        } else {
          console.log("setState no se llamÃ³");
        }
      })
      .catch(()=> {
        setState({
          data: null,
          loading: false,
          error: 'No se pudo cargar la info'
        })
      })
  }, [url]);

  //Retormamos el estado correspondiente al objeto con lo valores resultados del fetch
  return state;
};
