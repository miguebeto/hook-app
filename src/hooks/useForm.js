import { useState } from "react";

export const useForm = (initialState = {}) => {
  //creamos el custom hook con el estado, la funcion seteadora y el estadoinicial que se recibe por prop y por defecto es un objeto 
  const [values, setValues] = useState(initialState);

  //establece como valor inicial del estado el valo que nos pasan por parÃ¡metro
  const reset = () => {
    setValues( initialState );
  }

  //recibe el target del evento onchange y setea el valor del estado con un nuevo objeto que recibe el valor anterior y agrega el name: value del input
  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [target.name]: target.value,
    });
  };

  //retomamos el estado, el manejador de los eventos onChange y el reset
  return [ values, handleInputChange, reset ]

};
