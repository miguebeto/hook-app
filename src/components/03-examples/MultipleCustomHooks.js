import React from "react";
import { useCounterQuotes } from "../../hooks/useCounterQuotes";
import { useFetch } from "../../hooks/useFetch";
import "../02-useEffect/effect.css";

export const MultipleCustomHooks = () => {

  //traemos el custom hook con el estado y la funcion que incrementa en 1 el valor del estado
  const { counter, increment } = useCounterQuotes(1);

  //traemos elcustom hook del fetch pasandole la url de la api y desestructurando la data y el loading
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${ counter }`
  );

  //Desestructuramos author de y quote de la data en caso de que tenga valor
  const { author, quote } = !!data && data[0];

  return (
    <div>
      <h1>Breaking Bad Quotes</h1>
      <hr />
      {/* si loading es true mandamos el texto loading de contrario mandamos el quote y el author */}
      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-2"> {quote} </p>
          <footer className="blockquote-footer"> {author} </footer>
        </blockquote>
      )}
      {/* presionamos el boton para icrementar el valor del id correspodiente a cada quote  */}
      <button onClick={ increment } className="btn btn-primary">Siguiente quote</button>
    </div>
  );
};
