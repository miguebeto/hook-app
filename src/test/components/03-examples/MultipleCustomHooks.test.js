import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks }  from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

//simula el hooks evitando las llamadas al useFetch para no alterarlo
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Pruebas en el componente <MultipleCustomHooks />', ()=> {
    //simulamos el retorno del hook con la valores que manda al estado inicial
    useCounter.mockReturnValue({
        counter: 10,
        increment: () => {}
    })

    test('Debe mantenerse el componente sin cambios en el renderizado', ()=>{

       //simulamos el retorno de la data por defecto en el hook
       useFetch.mockReturnValue({
           data: null,
           loading: true,
           error: null
       })

       //debe retornar el componente sin cambios en el render
       const wrapper = shallow(<MultipleCustomHooks />);
       expect(wrapper).toMatchSnapshot();
    });

    test('Debe mostrar la información', ()=>{

         //simulamos el retorno de la data traida de la api
       useFetch.mockReturnValue({
        data: [{
            author: 'Miguel',
            quote: 'Hola Mundo'
        }],
        loading: false,
        error: null
    });
        //guardamos el componente en la variable wrapper
        const wrapper = shallow(<MultipleCustomHooks />);

        console.log(wrapper.html());

        //comprueba que el loadin no se ejecutó verificando su className
        expect( wrapper.find('.alert').exists()).toBe(false);

        //comprueba que el className del párrafo que contiene .mb-0 tenga el texto 'Hola mundo' que se pasó en la simulación con mock
        expect( wrapper.find('.mb-2').text().trim()).toBe('Hola Mundo');

        //comprueba que el valor de la etiqueta footer es igual al valor que pasamo en el author de la simulación
        expect( wrapper.find("footer").text().trim()).toBe('Miguel');


        
    });
});