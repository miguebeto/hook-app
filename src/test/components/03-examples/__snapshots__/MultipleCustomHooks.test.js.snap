// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <MultipleCustomHooks /> Debe mantenerse el componente sin cambios en el renderizado 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
